### PR TITLE
[reconfig] Add certificate creation trait

### PR DESCRIPTION
**Pull request by lxfind**
At the end of epoch, the checkpoint builder would need to create an authority aggregator on the fly from the system state, and use it to construct a certificate for the advance epoch transaction. Currently this is done explicitly using the network client. However that makes testing difficult. If we want to unit test individual components without having to start a network node and Narwhal, we need to use a local client.
This PR hides the use of network vs local client through a trait object, and provide different implementations for it.

### CODE REVIEW

File: crates/sui-core/src/test_utils.rs [hunk 1]
```diff
@@ -149,3 +164,113 @@ pub fn dummy_transaction_effects(tx: &Transaction) -> TransactionEffects {
         dependencies: Vec::new(),
     }
 }
+
+async fn init_genesis(
+    committee_size: usize,
+    mut genesis_objects: Vec<Object>,
+) -> (
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-core/src/test_utils.rs:171]: oh this is moved code...even still i wonder if it could be eliminated.

**Review by bmwill — COMMENTED** [review]
Seems reasonable. One thing i'm not quite sure of is why do validators need to use this network interface and drive quorum on txns themselves? We probably also want to not use the grpc interface for validator to validator communication due to it not being encrypted or authenticated

**Comment by lxfind** [comment]
> One thing i'm not quite sure of is why do validators need to use this network interface and drive quorum on txns themselves?

What are the alternatives? Talking to a fullnode?

**Comment by lxfind** [comment]
btw I think down the road, when we support storing/executing finalized transactions without certs, we could then get away without having to do this quorum dance

**Comment by bmwill** [comment]
> What are the alternatives? Talking to a fullnode?

I suppose i just don't understand which transactions they would even need to do this for.

**Comment by lxfind** [comment]
Oh it's this last transaction of the epoch that advances epoch on-chain and distribute gas rewards. It's a system transaction and each validator builds it independently, but still need to collect signatures from each other to form a cert.

File: crates/sui-core/src/test_utils.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lxfind) [reply crates/sui-core/src/test_utils.rs:171]: I think we could. But the cleanups are probably non-trivial. Let me follow up in separate PRs.

**Comment by bmwill** [comment]
>Oh it's this last transaction of the epoch that advances epoch on-chain and distribute gas rewards. It's a system transaction and each validator builds it independently, but still need to collect signatures from each other to form a cert.

Is this necessary? this seems like an unnecessary increase to latency. I would have expected that at the end of an epoch there are some signals from narwhal that we use to signal the ending of the epoch and that we need to form the final checkpoint of the epoch. At that point all validators would independently construct and execute this system transaction without any input from each other. Once that finishes they would finish crafting the final checkpoint, doing the narwhal dance to finalize that checkpoint which will include this system transaction and its effects.

**Comment by lxfind** [comment]
I agree with you. However in the current implementation we require a transaction certificate ever where for any transaction that's committed to the system. Once we do the finalized transaction work where we support executing and storing finalized transactions in raw forms we wouldn't need this anymore. But in today's code base we don't have a choice but to create a cert on the fly.

**Review by bmwill — APPROVED** [review]
It sounds like we're going to need to do a bit of iteration to get to the ideal solution so approving this to unblock

File: crates/sui/tests/reconfiguration_tests.rs [hunk 2]
```diff
@@ -1,78 +1,95 @@
...
+        .create_advance_epoch_tx_cert(
+            1,
+            &GasCostSummary::new(0, 0, 0),
+            Duration::from_secs(15),
+            certifier,
+        )
         .await;
     // Since we are only running the task on one validator, it will never get a quorum and hence
     // never succeed.
     assert!(failing_task.is_err());
```
Comment (longbowlu) [inline crates/sui/tests/reconfiguration_tests.rs:71]: check it's `SuiError::InvalidSystemTransaction` or a variant of it?

**Review by longbowlu — COMMENTED** [review]
One thing I don't see we have to do, at least from this PR, is passing `TransactionCertifier` to `CheckpointService`. In a unit test setup where we focus on testing stuff like `create_advance_epoch_tx_cert`, we can pass the certifier when calling it. In non-unit test code path, we could init `NetworkTransactionCertifier` in e.g. `fn augment_epoch_last_checkpoint` and avoid passing it all the way from `CheckpointService`

**Comment by lxfind** [comment]
> One thing I don't see we have to do, at least from this PR, is passing TransactionCertifier to CheckpointService. In a unit test setup where we focus on testing stuff like create_advance_epoch_tx_cert, we can pass the certifier when calling it. In non-unit test code path, we could init NetworkTransactionCertifier in e.g. fn augment_epoch_last_checkpoint and avoid passing it all the way from CheckpointService

If create_advance_epoch_tx_cert is the only function we want to test, then yes.
But let's say we want to test that the last checkpoint properly augments itself with the advance epoch tx. How would you test it with a local client setup?

File: crates/sui-core/src/authority_aggregator.rs [hunk 3]
```diff
@@ -404,6 +406,88 @@ impl AuthorityAggregator<NetworkAuthorityClient> {
...
+    async fn create_certificate(
+        &self,
+        transaction: &VerifiedTransaction,
+        self_state: &AuthorityState,
+        timeout: Duration,
+    ) -> anyhow::Result<VerifiedCertificate> {
+        let net = AuthorityAggregator::new_from_system_state(
+            &self_state.db(),
+            self_state.committee_store(),
+            &Registry::new(),
```
[the remark this answers is not in the document]
Comment (lxfind) [reply crates/sui-core/src/authority_aggregator.rs:457]: SafeClient will always be there.
The only thing we are missing in this case is the metrics. I don't mind stuffing a registry somewhere though.

**Pull request merged**